File: Leetcode/86-Partition-List/PartitionList.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* partition(ListNode* head, int x) {
        ListNode* cur = head;

        ListNode* leftDummy = new ListNode(0);
        ListNode* leftCur = leftDummy;

        ListNode* rightDummy = new ListNode(0);
        ListNode* rightCur = rightDummy;
        while (cur != NULL) {
            if (cur->val < x) {
                leftCur->next = cur;
                cur = cur->next;
                leftCur = leftCur->next;
                leftCur->next = NULL;
            } else {
                rightCur->next = cur;
                cur = cur->next;
                rightCur = rightCur->next;
                rightCur->next = NULL;
            }
        }
        leftCur->next = rightDummy->next;
        return leftDummy->next;
    }
};

File: Leetcode/86-Partition-List/PartitionList.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func partition(head *ListNode, x int) *ListNode {
	cur := head

	leftDummy := &ListNode{Val: 0}
	leftCur := leftDummy

	rightDummy := &ListNode{Val: 0}
	rightCur := rightDummy

	for cur != nil {
		if cur.Val < x {
			leftCur.Next = cur
			cur = cur.Next
			leftCur = leftCur.Next
			leftCur.Next = nil
		} else {
			rightCur.Next = cur
			cur = cur.Next
			rightCur = rightCur.Next
			rightCur.Next = nil
		}
	}
	leftCur.Next = rightDummy.Next
	return leftDummy.Next
}
